update discovery.rs

hawkw opened the pull request:
This seems to be correct. There's one remaining type error in `control/mod.rs`, but I'm pretty sure it requires changes in _that_ file, and I didn't want to touch anything outside of `discovery.rs`. @carllerche, WDYT?

hawkw committed:
Fix Discovery::poll_reconnect

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
discovery compiles

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
Reduce ugliness with type aliases

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
discovery should *actually* work now

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

carllerche commented on proxy/src/control/discovery.rs:
```diff
@@ -29,6 +30,11 @@ pub type ClientBody = ::tower_grpc::client::codec::EncodingBody<
+type UpdateRsp<F> =
+    grpc::client::server_streaming::ResponseFuture<PbUpdate, F>;
+
+type UpdateStream<E> = Stream<Item=PbUpdate, Error=E>;
```
I don't see this used anywhere in the diff?

hawkw replied:
Ah, yeah this is dead & should be removed.

carllerche commented on proxy/src/control/discovery.rs:
```diff
@@ -66,14 +78,31 @@ pub struct DiscoveryWork<T: HttpService> {
 
+enum UpdateRx<F, S>
+where
+    F: Future<Item=grpc::Response<S>>,
```
Are all of these bounds required at the struct / enum level?

carllerche commented on proxy/src/control/discovery.rs:
```diff
@@ -187,10 +219,17 @@ impl Background {
+    S: Stream<Item=PbUpdate, Error=E>,
+    <UpdateRsp<T::Future> as Future>::Error: fmt::Debug,
+    UpdateRx<UpdateRsp<T::Future>, S>: Stream<Item=PbUpdate>,
+    <UpdateRx<UpdateRsp<T::Future>, S> as Stream>::Error: fmt::Debug,
```
I haven't really figured out why all of these where bounds are required. Is there anyway to trim this down?

hawkw replied:
@carllerche this is my best attempt at reducing the number of bounds; I can try to see if some of them can be removed, but it used to look [much worse](https://github.com/carllerche/conduit/pull/2/commits/5138f0392a391978484d4d3d7fb638f717db8232?diff=unified#diff-3d7563188396c0392342b192778deb15R209). I'll see if it can be reduced further, but I doubt it.

carllerche reviewed (commented):
As far as I can tell, it looks mostly good.

My main request would be to try to simplify the bounds as much as possible. It tends to add noise and makes it harder to figure out what is going on.

It looks like there are a number of struct / enum level bounds, but I couldn't tell how they were required.

hawkw committed:
Remove bounds on structs/enums

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
Simplify where bounds

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw commented:
@carllerche I think 4e2f1cff4672989c9fbbe6f36c003747d535e48b is the smallest set of `where` bounds on `DiscoverWork` so far that still compiles. I've removed the bounds at the struct/enum level --- I think they were required at one point, but I must have made some other changes allowing them to be removed.

hawkw committed:
document UpdateRx

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
Merge branch 'cl/update-deps' of https://github.com/carllerche/conduit into cl/update-deps

hawkw committed:
integrate @carllerche's changes to make type bounds less horrendous

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
Depend on fixed revision of tower-balance for now

We can work on integrating tower-rs/tower#35 later, but that seems out of scope for this changeset.

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
Depend on tower-rs/tower-h2#18

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
start updating bounds on Outbound

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
fix compile errors in outbound

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
LiftService is on master now!

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
use new lb constructors

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
everything compiles now!!!

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
Remove unused imports

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
Remove use of deprecated methods

hawkw committed:
Update tests with renamed tower-h2 types

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
remove unnecessary bounds

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
Update Cargo lockfile and proxy deps SHAs

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
Remove commented-out code

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
Turn warnings back on

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw committed:
Remove unused imports

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

hawkw commented:
Closing this in favour of runconduit/conduit#215
